// @toruslabs/fetch-node-details/dist/nodeDetailManager.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _web3EthContract = _interopRequireDefault(require("web3-eth-contract"));

var _web3Utils = require("web3-utils");

var _abi = require("./abi.json");

var NodeDetailManager =
/*#__PURE__*/
function () {
  function NodeDetailManager() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        _ref$network = _ref.network,
        network = _ref$network === void 0 ? "mainnet" : _ref$network,
        _ref$proxyAddress = _ref.proxyAddress,
        proxyAddress = _ref$proxyAddress === void 0 ? "0x638646503746d5456209e33a2ff5e3226d698bea" : _ref$proxyAddress;

    (0, _classCallCheck2["default"])(this, NodeDetailManager);
    (0, _defineProperty2["default"])(this, "_currentEpoch", "13");
    (0, _defineProperty2["default"])(this, "_torusNodeEndpoints", ["https://binance-main-13.torusnode.com/jrpc", "https://waseda-main-13.torusnode.com/jrpc", "https://vgr-main-13.torusnode.com/jrpc", "https://torus-main-13.torusnode.com/jrpc", "https://etc-main-13.torusnode.com/jrpc"]);
    (0, _defineProperty2["default"])(this, "_torusNodePub", [{
      X: "801e99d5c6befe4286de6f22b086508c02ae5730afabf15f316242238fa8c832",
      Y: "a4791de3ebd17197d09d7a52d417994e7fdab17f8de002fc139c90dfdebfe7e1"
    }, {
      X: "f4034ed9b11a3573d496b09823f69efed1213bc903f0b520652cb1b80f72e5c3",
      Y: "c1619b0d594db4bb289b2ec18c029e777b1e04aa0493f7318d14a68e89ff0005"
    }, {
      X: "10aeb138801ef2e189e21a016e913fc2c2d40463e90e09687c6cef011e64048d",
      Y: "1dbc9e567db833376cac16ad07b402d44458befb26125e98446270d41e759a99"
    }, {
      X: "1a864642b3e612615c0db4999cc89e123b7ab6bafe58b1692d46b64ba27508f6",
      Y: "92919c5269b22818a9e62d246df282b7847782e46f94d775d8d20fe967fb5cab"
    }, {
      X: "9b64f7f8db105bb5a9301ac7ab20958e6d71f4a231f0d29b431712dd61d767a7",
      Y: "ee7de7de555aea2e586c9b40e68fab778742c0a8871e8b8f88a97063822764b4"
    }]);
    (0, _defineProperty2["default"])(this, "_torusIndexes", [1, 2, 3, 4, 5]);
    var url;

    try {
      var localUrl = new URL(network);
      url = localUrl.href;
    } catch (_) {
      url = "https://api.infura.io/v1/jsonrpc/".concat(network);
    }

    try {
      _web3EthContract["default"].setProvider(url);

      this.nodeListContract = new _web3EthContract["default"](_abi.abi, proxyAddress);
      this.nodeListAddress = proxyAddress;
      this.updated = false;
    } catch (error) {
      throw new Error(error);
    }
  }

  (0, _createClass2["default"])(NodeDetailManager, [{
    key: "getCurrentEpoch",
    value: function getCurrentEpoch() {
      return this.nodeListContract.methods.currentEpoch().call();
    }
  }, {
    key: "getEpochInfo",
    value: function getEpochInfo(epoch) {
      return this.nodeListContract.methods.getEpochInfo(epoch).call();
    }
  }, {
    key: "getNodeEndpoint",
    value: function getNodeEndpoint(nodeEthAddress) {
      return this.nodeListContract.methods.getNodeDetails(nodeEthAddress).call();
    }
  }, {
    key: "getNodeDetails",
    value: function getNodeDetails() {
      var _this = this;

      var skip = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
      return new Promise(function (resolve, reject) {
        if (skip) return resolve(_this._nodeDetails);
        if (_this.updated) return resolve(_this._nodeDetails);

        _this.getCurrentEpoch().then(function (latestEpoch) {
          _this._currentEpoch = latestEpoch;
          return _this.getEpochInfo(latestEpoch);
        }).then(function (latestEpochInfo) {
          var nodeEndpointRequests = [];
          var indexes = latestEpochInfo.nodeList.map(function (_, pos) {
            return pos + 1;
          });
          _this._torusIndexes = indexes;
          latestEpochInfo.nodeList.map(function (nodeEthAddress) {
            nodeEndpointRequests.push(_this.getNodeEndpoint(nodeEthAddress)["catch"](function (_) {}));
          });
          return Promise.all(nodeEndpointRequests);
        }).then(function (nodeEndPoints) {
          var updatedEndpoints = [];
          var updatedNodePub = [];

          for (var index = 0; index < nodeEndPoints.length; index++) {
            var endPointElement = nodeEndPoints[index];
            var endpoint = "https://".concat(endPointElement.declaredIp.split(":")[0], "/jrpc");
            updatedEndpoints.push(endpoint);
            updatedNodePub.push({
              X: (0, _web3Utils.toHex)(endPointElement.pubKx).replace("0x", ""),
              Y: (0, _web3Utils.toHex)(endPointElement.pubKy).replace("0x", "")
            });
          }

          _this._torusNodeEndpoints = updatedEndpoints;
          _this._torusNodePub = updatedNodePub;
          _this.updated = true;
          resolve(_this._nodeDetails);
        })["catch"](function (_) {
          return resolve(_this._nodeDetails);
        });
      });
    }
  }, {
    key: "_nodeDetails",
    get: function get() {
      return {
        currentEpoch: this._currentEpoch,
        nodeListAddress: this.nodeListAddress,
        torusNodeEndpoints: this._torusNodeEndpoints,
        torusNodePub: this._torusNodePub,
        torusIndexes: this._torusIndexes,
        updated: this.updated
      };
    }
  }]);
  return NodeDetailManager;
}();

var _default = NodeDetailManager;
exports["default"] = _default;
